grammar/cypher.xml

```diff
@@ -96,68 +96,99 @@
 
   <production name="SinglePartQuery">
     <alt>
-      <non-terminal ref="ReadOnlyEnd"/>
-      <non-terminal ref="ReadUpdateEnd"/>
-      <non-terminal ref="UpdatingEnd"/>
+      <seq>
+        <non-terminal ref="ReadOnlyQueryPartBody"/><!-- whitespace that follows already consumed in rule -->
+        <non-terminal ref="Return"/>
+      </seq>
+      <seq>
+        <non-terminal ref="ReadUpdateQueryPartBody"/>
+        <opt>&WS; <non-terminal ref="Return"/></opt>
+      </seq>
+      <seq>
+        <non-terminal ref="UpdateOnlyQueryPartBody"/>
+        <opt>&WS; <non-terminal ref="Return"/></opt>
+      </seq>
     </alt>
   </production>
 
-  <production name="ReadOnlyEnd" rr:inline="true">
-    <non-terminal ref="ReadPart"/>
-    <non-terminal ref="Return"/>
+  <production name="ReadOnlyQueryPartBody" rr:inline="true">
+    <!-- whitespaces consumed at end of rule -->
+    <repeat>
+      <non-terminal ref="ReadingClause"/>
+      &WS;
+    </repeat>
   </production>
 
-  <production name="ReadUpdateEnd" rr:inline="true">
+  <production name="ReadUpdateQueryPartBody" rr:inline="true">
+    <!-- no whitespaces consumed at begin/end of rule -->
     <non-terminal ref="ReadingClause"/>
     <repeat>
       &WS; <non-terminal ref="ReadingClause"/>
     </repeat>
     <repeat min="1">
       &WS; <non-terminal ref="UpdatingClause"/>
     </repeat>
-    <opt>&WS; <non-terminal ref="Return"/></opt>
   </production>
 
-  <production name="UpdatingEnd" rr:inline="true">
+  <production name="UpdateOnlyQueryPartBody" rr:inline="true">
```

Comment:
I think this can be simplified even further, by removing the class `UpdatingStartClause` altogether.